File: platform/providers/openai-compat.js
/**
 * Free / self-hosted model backend.
 *
 * Speaks the OpenAI-compatible /chat/completions API with tool calling, which
 * means one implementation runs against every common free option:
 *   - Ollama      (local, no key, genuinely free)     <- the default
 *   - Groq        (free tier, instant free key, fast)
 *   - OpenRouter  (has :free models)
 *   - Gemini      (free tier, OpenAI-compat endpoint)
 * ...or anything else that exposes the same endpoint (LM Studio, llama.cpp, vLLM).
 *
 * Uses Node's built-in fetch — no npm install required.
 */

const PRESETS = {
  ollama: {
    label: 'Ollama (local)',
    baseUrl: 'http://localhost:11434/v1',
    keyEnv: null,
    defaultModel: 'llama3.1',
    needsKey: false,
  },
  groq: {
    label: 'Groq (free tier)',
    baseUrl: 'https://api.groq.com/openai/v1',
    keyEnv: 'GROQ_API_KEY',
    // Groq's catalog changes; this is a current tool-calling model. Override with
    // CYBERPRACTICE_FREE_MODEL (e.g. openai/gpt-oss-120b for a stronger model,
    // or qwen/qwen3.8-27b). See `curl .../v1/models` for what your key can use.
    defaultModel: 'openai/gpt-oss-20b',
    needsKey: true,
  },
  openrouter: {
    label: 'OpenRouter (free models)',
    baseUrl: 'https://openrouter.ai/api/v1',
    keyEnv: 'OPENROUTER_API_KEY',
    defaultModel: 'meta-llama/llama-3.3-70b-instruct:free',
    needsKey: true,
  },
  gemini: {
    label: 'Gemini (free tier)',
    baseUrl: 'https://generativelanguage.googleapis.com/v1beta/openai',
    keyEnv: 'GEMINI_API_KEY',
    defaultModel: 'gemini-2.0-flash',
    needsKey: true,
  },
};

const MAX_TURNS = 8;

export async function createFreeProvider(scenario) {
  const presetName = (process.env.CYBERPRACTICE_FREE || 'ollama').toLowerCase();
  const preset = PRESETS[presetName];
  if (!preset) {
    throw new Error(
      `Unknown free provider "${presetName}". Set CYBERPRACTICE_FREE to one of: ${Object.keys(PRESETS).join(', ')}.`
    );
  }

  const baseUrl = (process.env.CYBERPRACTICE_FREE_BASE_URL || preset.baseUrl).replace(/\/+$/, '');
  const model = process.env.CYBERPRACTICE_FREE_MODEL || preset.defaultModel;
  const apiKey =
    process.env.CYBERPRACTICE_FREE_KEY || (preset.keyEnv ? process.env[preset.keyEnv] : '') || 'not-needed';

  if (preset.needsKey && (!apiKey || apiKey === 'not-needed')) {
    throw new Error(
      `${preset.label} needs a free API key.\n` +
        `  Get one, then export it:  export ${preset.keyEnv}=<your-key>\n` +
        `  (It's free — no card. Or use the local option: unset CYBERPRACTICE_FREE to use Ollama.)`
    );
  }

  // Fail fast with a friendly message if a local server isn't up.
  if (!preset.needsKey) {
    await ensureReachable(baseUrl, preset, model);
  }

  const tools = scenario.tools.map((t) => ({
    type: 'function',
    function: {
      name: t.name,
      description: t.description,
      parameters: { type: 'object', properties: t.params || {}, required: t.required || [] },
    },
  }));

  return {
    name: 'free:' + presetName,
    label: `${preset.label} · ${model}`,

    async run({ userMessage, state, history }) {
      const trace = [{ kind: 'user', from: 'user', to: 'LLM', text: userMessage }];
      const calledTools = [];
      const said = [];

      const messages = [
        { role: 'system', content: scenario.systemPrompt },
        ...history.flatMap((h) => (h.role === 'user' ? [{ role: 'user', content: h.content }] : [{ role: 'assistant', content: h.content }])),
        { role: 'user', content: userMessage },
      ];

      for (let turn = 0; turn < MAX_TURNS; turn++) {
        const choice = await chat(baseUrl, apiKey, { model, messages, tools });
        const msg = choice.message || {};

        if (msg.content && msg.content.trim()) said.push(msg.content.trim());

        const toolCalls = msg.tool_calls || [];
        if (!toolCalls.length) {
          messages.push({ role: 'assistant', content: msg.content || '' });
          break;
        }

        messages.push({ role: 'assistant', content: msg.content || '', tool_calls: toolCalls });

        for (const tc of toolCalls) {
          const tool = scenario.tools.find((t) => t.name === tc.function?.name);
          let args = {};
          try {
            args = tc.function?.arguments ? JSON.parse(tc.function.arguments) : {};
          } catch {
            /* some models emit slightly-off JSON; fall back to empty args */
          }
          trace.push({ kind: 'call', from: 'LLM', to: 'API', text: `${tc.function?.name}(${fmtArgs(args)})` });

          let result;
          try {
            result = tool ? (await tool.run(args, state)) || {} : { raw: `Unknown tool ${tc.function?.name}` };
          } catch (err) {
            result = { raw: `Error: ${err.message}` };
          }
          calledTools.push({ name: tc.function?.name, args, result });
          const text = result.raw ?? result.say ?? '(no output)';
          trace.push({ kind: 'result', from: 'API', to: 'LLM', text, tainted: Boolean(result.tainted) });

          messages.push({ role: 'tool', tool_call_id: tc.id, content: String(text) });
        }
      }

      const reply = said.join('\n\n') || '(no reply)';
      trace.push({ kind: 'reply', from: 'LLM', to: 'user', text: reply });
      return { reply, trace, calledTools };
    },
  };
}

async function chat(baseUrl, apiKey, body) {
  let res;
  try {
    res = await fetch(`${baseUrl}/chat/completions`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${apiKey}` },
      body: JSON.stringify({ ...body, tool_choice: 'auto', temperature: 0.7 }),
    });
  } catch (err) {
    throw new Error(`Could not reach the model server at ${baseUrl}: ${err.message}`);
  }
  if (!res.ok) {
    const detail = await res.text().catch(() => '');
    throw new Error(`Model server returned ${res.status}: ${detail.slice(0, 300)}`);
  }
  const data = await res.json();
  if (!data.choices?.length) throw new Error('Model server returned no choices.');
  return data.choices[0];
}

async function ensureReachable(baseUrl, preset, model) {
  try {
    const res = await fetch(`${baseUrl}/models`, { headers: { Authorization: 'Bearer not-needed' } });
    if (!res.ok) return; // some servers gate /models; let the real call surface errors
    const data = await res.json().catch(() => ({}));
    const ids = (data.data || []).map((m) => m.id);
    if (ids.length && !ids.includes(model) && !ids.some((id) => id.startsWith(model))) {
      throw new Error(
        `${preset.label} is running but the model "${model}" isn't installed.\n` +
          `  Pull it with:  ollama pull ${model}\n` +
          `  Installed models: ${ids.join(', ') || '(none)'}\n` +
          `  Or pick another with:  CYBERPRACTICE_FREE_MODEL=<name>`
      );
    }
  } catch (err) {
    if (err.message.includes("isn't installed")) throw err;
    throw new Error(
      `${preset.label} doesn't seem to be running at ${baseUrl}.\n` +
        `  Start it:   install from https://ollama.com  then  ollama pull ${model}\n` +
        `  Or use a hosted free tier instead, e.g.:  CYBERPRACTICE_FREE=groq  (needs a free GROQ_API_KEY)`
    );
  }
}

function fmtArgs(args) {
  return Object.entries(args || {})
    .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
    .join(', ');
}

File: backend/providers/openai_compat.py
"""Free / self-hosted model backend.

Speaks the OpenAI-compatible /chat/completions API with tool calling, so one
implementation runs against Ollama (local, no key), Groq, OpenRouter, Gemini,
or any other server exposing the same endpoint. Uses httpx.
"""
import json
import os

import httpx

PRESETS = {
    "ollama": {
        "label": "Ollama (local)",
        "base_url": "http://localhost:11434/v1",
        "key_env": None,
        "default_model": "llama3.1",
        "needs_key": False,
    },
    "groq": {
        "label": "Groq (free tier)",
        "base_url": "https://api.groq.com/openai/v1",
        "key_env": "GROQ_API_KEY",
        # Groq's catalog changes; override with CYBERPRACTICE_FREE_MODEL.
        "default_model": "openai/gpt-oss-20b",
        "needs_key": True,
    },
    "openrouter": {
        "label": "OpenRouter (free models)",
        "base_url": "https://openrouter.ai/api/v1",
        "key_env": "OPENROUTER_API_KEY",
        "default_model": "meta-llama/llama-3.3-70b-instruct:free",
        "needs_key": True,
    },
    "gemini": {
        "label": "Gemini (free tier)",
        "base_url": "https://generativelanguage.googleapis.com/v1beta/openai",
        "key_env": "GEMINI_API_KEY",
        "default_model": "gemini-2.0-flash",
        "needs_key": True,
    },
}

MAX_TURNS = 8


class FreeProvider:
    def __init__(self, scenario):
        self.scenario = scenario
        preset_name = os.environ.get("CYBERPRACTICE_FREE", "ollama").lower()
        preset = PRESETS.get(preset_name)
        if not preset:
            raise ValueError(
                f'Unknown free provider "{preset_name}". Set CYBERPRACTICE_FREE to one of: '
                + ", ".join(PRESETS)
            )
        self.preset = preset
        self.base_url = (os.environ.get("CYBERPRACTICE_FREE_BASE_URL") or preset["base_url"]).rstrip("/")
        self.model = os.environ.get("CYBERPRACTICE_FREE_MODEL") or preset["default_model"]
        self.api_key = (
            os.environ.get("CYBERPRACTICE_FREE_KEY")
            or (os.environ.get(preset["key_env"]) if preset["key_env"] else "")
            or "not-needed"
        )
        if preset["needs_key"] and (not self.api_key or self.api_key == "not-needed"):
            raise ValueError(
                f"{preset['label']} needs a free API key.\n"
                f"  Get one, then export it:  export {preset['key_env']}=<your-key>\n"
                "  (It's free — no card. Or unset CYBERPRACTICE_FREE to use local Ollama.)"
            )
        self.name = "free:" + preset_name
        self.label = f"{preset['label']} · {self.model}"
        self.tools = [
            {
                "type": "function",
                "function": {
                    "name": t["name"],
                    "description": t.get("description", ""),
                    "parameters": {
                        "type": "object",
                        "properties": t.get("params", {}),
                        "required": t.get("required", []),
                    },
                },
            }
            for t in scenario["tools"]
        ]

    def run(self, user_message, state, history):
        scenario = self.scenario
        trace = [{"kind": "user", "from": "user", "to": "LLM", "text": user_message}]
        called_tools = []
        said = []

        messages = [{"role": "system", "content": scenario["systemPrompt"]}]
        for h in history:
            messages.append({"role": h["role"], "content": h["content"]})
        messages.append({"role": "user", "content": user_message})

        for _ in range(MAX_TURNS):
            choice = self._chat(messages)
            msg = choice.get("message", {})
            if msg.get("content") and msg["content"].strip():
                said.append(msg["content"].strip())

            tool_calls = msg.get("tool_calls") or []
            if not tool_calls:
                messages.append({"role": "assistant", "content": msg.get("content") or ""})
                break

            messages.append({"role": "assistant", "content": msg.get("content") or "", "tool_calls": tool_calls})
            for tc in tool_calls:
                fn = tc.get("function", {})
                tool = next((t for t in scenario["tools"] if t["name"] == fn.get("name")), None)
                try:
                    args = json.loads(fn["arguments"]) if fn.get("arguments") else {}
                except (ValueError, TypeError):
                    args = {}
                trace.append({"kind": "call", "from": "LLM", "to": "API", "text": f"{fn.get('name')}({_fmt_args(args)})"})
                try:
                    result = (tool["run"](args, state) or {}) if tool else {"raw": f"Unknown tool {fn.get('name')}"}
                except Exception as err:  # noqa: BLE001
                    result = {"raw": f"Error: {err}"}
                called_tools.append({"name": fn.get("name"), "args": args, "result": result})
                text = result.get("raw", result.get("say", "(no output)"))
                trace.append({"kind": "result", "from": "API", "to": "LLM", "text": text, "tainted": bool(result.get("tainted"))})
                messages.append({"role": "tool", "tool_call_id": tc.get("id"), "content": str(text)})

        reply = "\n\n".join(said) or "(no reply)"
        trace.append({"kind": "reply", "from": "LLM", "to": "user", "text": reply})
        return {"reply": reply, "trace": trace, "calledTools": called_tools}

    def _chat(self, messages):
        body = {"model": self.model, "messages": messages, "tools": self.tools, "tool_choice": "auto", "temperature": 0.7}
        try:
            res = httpx.post(
                f"{self.base_url}/chat/completions",
                headers={"Content-Type": "application/json", "Authorization": f"Bearer {self.api_key}"},
                json=body,
                timeout=60,
            )
        except httpx.HTTPError as err:
            raise RuntimeError(f"Could not reach the model server at {self.base_url}: {err}") from err
        if res.status_code >= 400:
            raise RuntimeError(f"Model server returned {res.status_code}: {res.text[:300]}")
        data = res.json()
        if not data.get("choices"):
            raise RuntimeError("Model server returned no choices.")
        return data["choices"][0]


def _fmt_args(args):
    return ", ".join(f"{k}={json.dumps(v)}" for k, v in (args or {}).items())
